File: app/src/main/java/io/mgba/UI/Views/GameInformationView.java
package io.mgba.UI.Views;

import android.content.Context;
import android.os.Bundle;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.flipboard.bottomsheet.BottomSheetLayout;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.mgba.Constants;
import io.mgba.Data.Database.Game;
import io.mgba.R;
import io.mgba.UI.Views.Interfaces.IGameInformationView;
import io.mgba.Utils.GlideUtils;
import io.mgba.Utils.GlideUtils.Colors;

public class GameInformationView implements IGameInformationView {

    private final Context context;
    @BindView(R.id.gameDescription)
    TextView gameDescription;
    @BindView(R.id.gameTitle)
    TextView gameTitle;
    @BindView(R.id.cover)
    ImageView cover;
    @BindView(R.id.bottomsheet_header)
    RelativeLayout bottomSheetHeader;
    @BindView(R.id.savestate_recyclerview)
    RecyclerView bottomSheetRecyclerview;
    @BindView(R.id.fab)
    FloatingActionButton fab;
    @BindView(R.id.sheet_container)
    CoordinatorLayout sheetContainer;
    @BindView(R.id.savestate_title)
    TextView savestateTitle;
    @BindView(R.id.no_content_image)
    ImageView noContentImage;
    @BindView(R.id.no_savestate_message)
    TextView noSavestateMessage;
    @BindView(R.id.no_savestates_container)
    RelativeLayout noSavestatesContainer;
    private View view;

    private Game currentShowing;

    public GameInformationView(Context ctx) {
        this.context = ctx;
    }

    @Override
    public View prepareView(BottomSheetLayout container, Game game) {
        currentShowing = game;

        view = LayoutInflater.from(context).inflate(R.layout.library_sheet_view, container, false);
        ButterKnife.bind(this, view);
        prepareView();

        return view;
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        outState.putParcelable(Constants.ARG_SHEET_CONTENT, currentShowing);
    }

    @Override
    public View prepareView(BottomSheetLayout container, Bundle inState) {
        Game game = (Game) inState.getParcelable(Constants.ARG_SHEET_CONTENT);

        return prepareView(container, game);
    }

    private void prepareView() {
        gameTitle.setText(currentShowing.getName());
        gameDescription.setText(currentShowing.getDescription());

        if(currentShowing.getCoverURL() != null) {
            GlideUtils.init(view, currentShowing.getCoverURL())
                      .setPlaceholders(R.drawable.placeholder, R.drawable.error)
                      .colorView(Colors.VIBRANT, Colors.DARK_MUTED, fab, savestateTitle)
                      .colorView(Colors.LIGHT_MUTED, Colors.LIGHT_VIBRANT, bottomSheetHeader, noSavestateMessage, noContentImage)
                      .colorView(Colors.LIGHT_VIBRANT, true, gameTitle)
                      .colorView(Colors.LIGHT_VIBRANT, false, gameDescription)
                      .build(cover);
        }
    }
}

File: app/src/main/java/io/mgba/ui/views/GameInformationView.kt
package io.mgba.ui.views

import android.content.Intent
import android.graphics.drawable.GradientDrawable
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import io.mgba.data.local.database.model.Game
import io.mgba.R
import io.mgba.emulation.EmulationActivity
import io.mgba.utilities.GlideUtils
import io.mgba.utilities.GlideUtils.Colors
import kotlinx.android.synthetic.main.game_information_view.*
import java.lang.RuntimeException

class GameInformationView : BottomSheetDialogFragment(), View.OnClickListener {

    private lateinit var game: Game

    override fun getTheme(): Int {
        return R.style.BottomSheetDialogTheme
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.game_information_view, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        if(arguments == null) throw RuntimeException("The bottomsheet could retrive the game_item!")

        arguments?.let {
            this.game = it.getParcelable(Constants.ARG_PLAY_GAME)!!
            prepareView()
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putParcelable(Constants.ARG_SHEET_CONTENT, game)
    }

    private fun prepareView() {
        game_title_bs.text = game.getName()
        game_desc_bs.text = game.description

        if (game_cover_bs != null && game.coverURL != null) {
            GlideUtils.init(this, game.coverURL!!)
                    .setPlaceholders(R.drawable.placeholder, R.drawable.error)
                    .colorView(Colors.VIBRANT, Colors.DARK_MUTED, game_play_bs, game_savetitle_bs)
                    .colorView(Colors.LIGHT_MUTED, Colors.LIGHT_VIBRANT, game_text_content_bs, game_image_content_bs)
                    .colorViewWithCustomBackground(Colors.LIGHT_MUTED, Colors.LIGHT_VIBRANT, game_header_bs)
                    .colorView(Colors.LIGHT_VIBRANT, true, game_title_bs)
                    .colorView(Colors.LIGHT_VIBRANT, false, game_desc_bs)
                    .build(game_cover_bs)
        } else {
            //set header color back to normal
            val background = game_header_bs.background

            if (background is GradientDrawable) {
                background.setColor(ContextCompat.getColor(context!!, R.color.colorPrimary))

            }
        }
    }

    override fun onClick(v: View) {
        val it = Intent(context, EmulationActivity::class.java)
        it.putExtra(Constants.ARG_PLAY_GAME, game)
        requireContext().startActivity(it)
    }

    companion object {

        fun show(game: Game): GameInformationView {
            val args = Bundle()
            args.putParcelable(Constants.ARG_PLAY_GAME, game)

            val frag = GameInformationView()
            frag.arguments = args
            return frag
        }
    }
}
